### PR TITLE
Log messages when starting / ending freeze

### PR DESCRIPTION
**Pull request by feerrenrut**
### Link to issue number:
As per https://github.com/nvaccess/nvda/pull/14289#issuecomment-1294403240

### Summary of the issue:
When NVDA freezes its behavior changes, this can break assumptions, and make inferring its behavior difficult.
The logs do not clearly indicate the start / end / level of a freeze.

### Description of user facing changes
None

### Description of development approach
- Refactor code to extract responsibilities to make loop conditions easier to reason about.
- Add logging for different levels of freeze, start and end.

### Testing strategy:
- System tests.
- Alpha testing.

### Known issues with pull request:
- It would be helpful to have clear names and a description of the implications of the different levels of freeze.

### Change log entries:
For Developers

```
- More detailed logging for NVDA freezes to aid debugging.
```

### Code Review Checklist:
- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] Security precautions taken.

### CODE REVIEW

File: source/watchdog.py [hunk 1]
```diff
@@ -104,48 +108,77 @@ def _isAlive():
...
-			if _isAlive():
-				break
+		else:
+			isAttemptingRecovery = True
+			waitForFreezeRecovery(waitedSince)
+
 		isAttemptingRecovery = False
 
+
+def waitForFreezeRecovery(waitedSince):
```
Comment (seanbudd) [inline source/watchdog.py:150]:
```suggestion
def waitForFreezeRecovery(waitedSince: float):
```

File: source/watchdog.py [hunk 2]
```diff
@@ -104,48 +108,77 @@ def _isAlive():
...
+	if log.isEnabledFor(log.DEBUGWARNING):
+		stacks = getFormattedStacksForAllThreads()
+		log.debugWarning(
+			f"Trying to recover from freeze after {_timer() - waitedSince} seconds."
+			f" Listing stacks for Python threads:\n{stacks}")
```
Comment (seanbudd) [inline source/watchdog.py:155]:
```suggestion
			f" Listing stacks for Python threads:\n{stacks}"
		)
```

File: source/watchdog.py [hunk 3]
```diff
@@ -104,48 +108,77 @@ def _isAlive():
...
+	else:
+		log.info(f"Starting freeze recovery.")
```
Comment (seanbudd) [inline source/watchdog.py:157]:
```suggestion
		log.info("Starting freeze recovery.")
```

**Review by seanbudd — COMMENTED** [review]
generally LGTM

File: source/watchdog.py [hunk 4]
```diff
@@ -21,22 +22,25 @@
...
-#: The minimum time to wait for the core to be alive.
-MIN_CORE_ALIVE_TIMEOUT=0.5
-#: How long to wait for the core to be alive under normal circumstances.
-#: This must be a multiple of MIN_CORE_ALIVE_TIMEOUT.
-NORMAL_CORE_ALIVE_TIMEOUT=10
-#: How long to wait between recovery attempts
+MIN_CORE_ALIVE_TIMEOUT = 0.5
+"""The minimum time (seconds) to wait for the core to be alive.
+"""
+NORMAL_CORE_ALIVE_TIMEOUT = 10
```
[the remark this answers is not in the document]
Comment (feerrenrut) [reply source/watchdog.py:28]: I'm actually not sure how true this comment is, the reasoning for this isn't given.

File: source/watchdog.py [hunk 5]
```diff
@@ -78,8 +78,20 @@ def alive():
 	windll.kernel32.ResetEvent(_cancelCallEvent)
 	# Set the timer so the watcher will take action in MIN_CORE_ALIVE_TIMEOUT
 	# if this function or asleep() isn't called.
-	windll.kernel32.SetWaitableTimer(_coreDeadTimer,
-		ctypes.byref(ctypes.wintypes.LARGE_INTEGER(-int(10000000 * MIN_CORE_ALIVE_TIMEOUT))),
+	SECOND_TO_100_NANOSECOND = 10000000  # nanosecond is 10^9, 10^7 is hundreds of nanoseconds
```
Comment (seanbudd) [inline source/watchdog.py:81]: I think this should be changed for readability

```suggestion
	SECOND_TO_100_NANOSECOND = 10**7  # nanosecond is 10^9, 10^7 is hundreds of nanoseconds
```

Or

```suggestion
	SECOND_TO_100_NANOSECOND = 10_000_000  # nanosecond is 10^9, 10^7 is hundreds of nanoseconds
```

**Review by seanbudd — APPROVED** [review]

File: source/watchdog.py (hunk 5 above)
Comment (feerrenrut) [reply source/watchdog.py:81]: Yep, good point.

**Comment by feerrenrut** [comment]
Last [system test failure](https://github.com/nvaccess/nvda/pull/14309#issuecomment-1299565005): 

```
Robot.startupShutdownNVDA.RestartsRobot
Error message:
Timed out waiting Welcome to NVDA to focus
```
Unlikely to be related to this change.

**Review by michaelDCurran — CHANGES_REQUESTED** [review]
I'd like to see this change split into separate commits in the following order:
* Switch to using performance counters
* Add missing logging
* Refactor code
As much as I don't want to create extra work here, I'm finding it hard to follow all the changes and therefore can not be confident to give my assurance that the change looks correct. Especially also because the primary point of the pr was to add missing logging, not to necessarily refactor.

**Comment by feerrenrut** [comment]
@michaelDCurran The changes have been split. The refactor is composed of several aspects, I also made these separate commits.

**Review by michaelDCurran — APPROVED** [review]
